internal/network/netns.go

```diff
@@ -21,94 +10,100 @@ import (
+	// Get a reference to the interface and make sure it's down
+	link, err := netlink.LinkByName(ifStr)
+	if err != nil {
+		return errors.Wrapf(err, "netlink.LinkByName(%s) failed", ifStr)
 	}
-
-	var a prot.NetworkAdapter
-	if err := json.Unmarshal([]byte(*cfgStr), &a); err != nil {
-		return err
+	if err := netlink.LinkSetDown(link); err != nil {
+		return errors.Wrapf(err, "netlink.LinkSetDown(%#v) failed", link)
 	}
 
-	if a.NatEnabled {
-		log.Infof("Configure %s in %d with: %s/%d gw=%s", *ifStr, *nspid, a.AllocatedIPAddress, a.HostIPPrefixLength, a.HostIPAddress)
-	} else {
-		log.Infof("Configure %s in %d with DHCP", *ifStr, *nspid)
+	// Move the interface to the new network namespace
+	if err := netlink.LinkSetNsPid(link, pid); err != nil {
+		return errors.Wrapf(err, "netlink.SetNsPid(%#v, %d) failed", link, pid)
 	}
+	return nil
+}
 
-	// Lock the OS Thread so we don't accidentally switch namespaces
+// DoInNetNS is a utility to run a function `work` inside of a specific network namespace
+// `ns`. This is accomplished by locking the current goroutines thread to prevent the goroutine
+// from being scheduled to a new thread during execution of `work`. The threads original network namespace
+// will be rejoined on exit.
+func DoInNetNS(ns netns.NsHandle, run func() error) error {
 	runtime.LockOSThread()
 	defer runtime.UnlockOSThread()
 
-	// Stash current network namespace away and make sure we enter it as we leave
-	log.Infof("Obtaining current namespace")
-	origNS, err := netns.Get()
+	origNs, err := netns.Get()
 	if err != nil {
-		return fmt.Errorf("netns.Get() failed: %v", err)
+		return errors.Wrap(err, "failed to get current network namespace")
 	}
-	defer origNS.Close()
-	log.Infof("Original namespace %v", origNS)
+	defer origNs.Close()
 
-	// Get a reference to the new network namespace
-	ns, err := netns.GetFromPid(*nspid)
-	if err != nil {
-		return fmt.Errorf("netns.GetFromPid(%d) failed: %v", *nspid, err)
+	if err := netns.Set(ns); err != nil {
+		return errors.Wrapf(err, "failed to set network namespace to %v", ns)
 	}
-	defer ns.Close()
-	log.Infof("New network namespace from PID %d is %v", *nspid, ns)
+	// Defer so we can re-enter the threads original netns on exit.
+	defer netns.Set(origNs)
 
-	// Get a reference to the interface and make sure it's down
-	link, err := netlink.LinkByName(*ifStr)
-	if err != nil {
-		return fmt.Errorf("netlink.LinkByName(%s) failed: %v", *ifStr, err)
+	return run()
+}
+
+// NetNSConfig moves a network interface into a network namespace and
+// configures it. The configuration specified by `adapterCfg` is expected to be
+// passed in as a JSON object (marshalled prot.NetworkAdapter). This function will log
+// it's output to `logFile`.
+//
+// This function MUST be used in tandem with `DoInNetNS` or some other means that ensures that the goroutine
+// executing this code stays on the same thread.
+func NetNSConfig(ifStr string, nsPid int, adapterCfg string, logFile string) error {
```

Comment:
Why not just pass in the prot.NetworkAdapter object here? The calling function marshals the object to a string and then this function just unmarshalls back

Reply:
Yea this is dumb, good call. Too much c+p